### ISSUE TITLE
Investigate Jacoco maven plugin for  test coverage

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#1173]
Base on #1167 , we can not always do workarounds and play with cobertura to make it happy.

Assignee: @MEZk 

Investigate:
0) open github issue against cobertura (it could be done base on workaround at #1167)
1) convert cobertura config to jacoco at pom.xml 
2) recheck how convenient HTML report of jacoco maven plugin

**Comment by MEZk** [comment]
There is a problem with JaCoCo  maven plugin configuration when I'm trying to  specify finer-grained thresholds for certain classes.

When we use Cobertura maven plugin we can specify finer-grained thresholds for certain classes using regular expressions.

For example:

```xml
<regex><pattern>.*.checkstyle.AnnotationUtility</pattern><branchRate>60</branchRate><lineRate>60</lineRate></regex>
```

In accordance with JaCoCo maven plugin [documentation for ```jacoco:check```](http://www.eclemma.org/jacoco/trunk/doc/check-mojo.html), we can also do this if we write the rule for every class with  ``` <include>``` tag in configuration for JaCoCo maven plugin (in pom.xml).

For example:

```xml
<configuration>
    <haltOnFailure>true</haltOnFailure>  // < ---- Halt the build if any of the checks fail
        <rules>
            <rule element="CLASS">
                <includes>
                   <include>com.puppycrawl.tools.checkstyle.AnnotationUtility</include>
                </includes>
                <limits>
                    <limit counter="BRANCH" value="COVEREDRATIO"><minimum>0.60</minimum></limit>
                    <limit counter="LINE" value="COVEREDRATIO"><minimum>0.60</minimum></limit>
               </limits>
            </rule>
        </rules>
</configuration>
```

But ```<include>``` tag doesn't work correctly and if any of the checks fail (coverage less than minimum requirements), haltOnFailure flag doesn't halt the build.

If we specify coverage thresholds for the whole project (for lines, branches, etc) it works fine! The problem appears only when we specify coverage thresholds for certain classes. 

I wrote about the problem at [JaCoCo support group](https://groups.google.com/forum/#!topic/jacoco/kdePtX4dJf8) and now waiting for answer.

**Comment by Godin** [comment]
@MEZk note that you don't need to write rule explicitely for each class - usage of wildcard characters (* and ?) is allowed in includes/excludes

**Comment by Godin** [comment]
and answering on jacoco/jacoco#326 - specify "element" as child node instead of attribute

**Comment by Godin** [comment]
Also if you have particular requirements for checks and can specify them, or if there is already experimental branch, then speaking as one of JaCoCo developers, I'll be happy to help with configuration to get JaCoCo adopted in such a great project as checkstyle. So don't hesitate to ask.

**Comment by MEZk** [comment]
@Godin thank you very much. I use your changes in my [sample project] (https://github.com/MEZk/JaCoCoTest) and it works fine.

@romani The problem with ```<include>``` tag within ```rule``` was resolved. I continue to convert cobertura config to jacoco at pom.xml for whole project.

**Comment by MEZk** [comment]
@romani 
Now we have another problem related to this issue.

I tried to add JaCoCo plugin into pom.xml:

````xml
     <plugin>
        <groupId>org.jacoco</groupId>
        <artifactId>jacoco-maven-plugin</artifactId>
        <version>0.6.2.201302030002</version>
        <executions>
          <execution>
            <id>prepare-agent</id>
            <goals>
              <goal>prepare-agent</goal>
            </goals>
          </execution>
          <execution>
            <id>report</id>
            <phase>site</phase>
            <goals>
              <goal>report</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
````

and in reporting section:

````xml
       <plugin>
        <groupId>org.jacoco</groupId>
        <artifactId>jacoco-maven-plugin</artifactId>
        <version>0.6.2.201302030002</version>
        <reportSets>
          <reportSet>
            <reports>
              <report>report</report>
            </reports>
          </reportSet>
        </reportSets>
      </plugin>
````
And I have this info-message:
```[INFO] --- jacoco-maven-plugin:0.6.2.201302030002:report (report) @ checkstyle ---
[INFO] Skipping JaCoCo execution due to missing execution data file```

According to your blog, you had the same [problem](http://roman-ivanov.blogspot.com/2014/06/skipping-jacoco-execution-problem-in.html). So I added ```${argLine} ``` into maven-surefire-plugin config:

````xml
<argLine>${argLine} -Duser.language=en -Duser.country=US -XX:-UseSplitVerifier</argLine>
````

but result remains  the same.

**Comment by MEZk** [comment]
@romani 
The problem was resolved.
maven-failsafe-plugin also requires inside configuration section:
````xml
<argLine>${argLine}</argLine>
````
In addition to this the location of jacoco.exec file should be declared directly for prepare-agent and report:

```xml
<configuration>
    <destFile>${basedir}/target/jacoco.exec</destFile>
</configuration>
````

````xml
<configuration>
    <dataFile>${basedir}/target/jacoco.exec</dataFile>
</configuration>
````

**Comment by MEZk** [comment]
@romani 
Coverage reports with excludes (for JaCoCo is the same as for Cobertura):
1) Cobertura
![cobertura](https://cloud.githubusercontent.com/assets/7242568/9028099/2a89bd7a-3976-11e5-845d-08ddb454dadf.png)
2) JaCoCo
![jacoco](https://cloud.githubusercontent.com/assets/7242568/9028100/2ab1ff56-3976-11e5-9498-ba0c6f5bde19.png)

[FULL REPORTS](http://mezk.github.io)

**Comment by romani** [comment]
1)
> ImportOrderCheck and/or InportOrderOption are reported uncovered by all versions (0.7.x) of JaCoCo.

please report issue against Jacoco and put there link to it. For now Cobertura looks better for our project, am I right ?

2) please create jacoco coverage confing in special profile ("-Pjacoco") of pom.xml . We will stay with cobertura till we enforce 100% on all classes (end of August). We could extend our Travis to test coverage by jacoco too, eventually the best tool survive.  On this point we are ok to close that issue.

**Comment by MEZk** [comment]
@romani 
I'll report issue against JaCoCo when you check [PR](https://github.com/checkstyle/checkstyle/pull/1549) with new JaCoCo profile of pom.xml.

>We could extend our Travis to test coverage by jacoco too, eventually the best tool survive. 

I think it will not make sense if JaCoCo's team does not fix bug related with fully uncovered class (in our case ImportOrderCheck and ImportOrderOption), because in this case we can not use haltOnFailure flag properly.

**Comment by Godin** [comment]
@MEZk let me quickly check PR #1549 about problem with class ImportOrderCheck

**Comment by MEZk** [comment]
@Godin It would be great, maybe I missed something.

**Comment by Godin** [comment]
@MEZk ImportOrderCheckTest uses PowerMock, which retransforms classes at runtime, so that what is actually executed isn't an original test and that's why JaCoCo isn't making an association to original source code. This is well-known difficulty" of usage JaCoCo with PowerMock. I believe that usage of offline instrumentation will help you.

**Comment by MEZk** [comment]
@Godin 
Offline instrumentation is ok for that purpose, thank you for help.

**Comment by Godin** [comment]
@MEZk there is an idea to provide pluggable mechanism for filtering for quite some time, but currently many synthetic constructs are not filtered, so options are
* explicitly cover them - this is definitely possible in cases as enum, private constructors, getters/setters, while might be difficult in case of switch-on-string
* exclude whole class

**Comment by MEZk** [comment]
@romani 
We have another problem with Power Mock and offline instrumentation. If ```mvn -P jacoco clean verify``` is used, 4 NPE will occur during test phase.
See [log](https://raw.githubusercontent.com/MEZk/MEZk.github.io/master/jacoco_run_log.txt).

@Godin
Could you please take a look?

**Comment by romani** [comment]
```
testUnableToReadSuppressions(com.puppycrawl.tools.checkstyle.filters.SuppressionsLoaderTest)  Time elapsed: 0.004 sec  <<< ERROR!
java.lang.NullPointerException: null
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.<init>(SuppressionsLoader.java:75)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadSuppressions(SuppressionsLoader.java:178)
```

That exception is happening from our code , please investigate a problem and provide workaround if possible.

**Comment by MEZk** [comment]
@romani 
I use reflection in tests to avoid 2 NPE:
```
  SuppressionsLoaderTest.testUnableToFindSuppressions:200 » NullPointer
  SuppressionsLoaderTest.testUnableToReadSuppressions:220 » NullPointer
```

We still have 2 NPE:
```
Tests in error: 
  ConfigurationLoaderTest.testLoadConfiguration_URISyntaxException:466 » NullPointer
  SuppressionsLoaderTest.testloadSuppressionsURISyntaxException:284 » NullPointer
```

testloadSuppressionsURISyntaxException  should invoke loadSuppressions which should throw CheckstyleException, but I can't imagine test input when URISyntaxException is happened [here](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionsLoader.java#L158).

testLoadConfiguration and  should invoke loadConfiguration which should throw URISyntaxException, the same, I can't imagine test input when CheckstyleException is thrown from loadConfiguration [here](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/ConfigurationLoader.java#L199).

In both cases the solution is to avoid using of Power Mock.

Exception testing with JUnit:
https://github.com/junit-team/junit/wiki/Exception-testing

**Comment by Godin** [comment]
@MEZk I'm currently in the way from vacation back home, so can take a look only very late today or tomorrow. But seems that this is not needed and issue not truly relate to JaCoCo. Isn't it?

**Comment by MEZk** [comment]
@Godin 
Yes, I think I'll solve that problem with NPE. Thank you for answer.

**Comment by romani** [comment]
>  but I can't imagine test input when URISyntaxException is happened

getResource method could throw such exception, I do not know how to make such use-case, most likely it is not possible if application is written correctly. Testing of that exception is dictated by 100% coverage demand, so I did such a testing with hack by PowerMock.

**Comment by romani** [comment]
your update for UTs is ok, lets finish with configuration, see comment in PR
